### DISCUSSION TITLE
How to test a scroll value of element for specific position?

### DISCUSSION THREAD

**Discussion by someApprentice** [cypress-io/cypress#26016] [Questions and Help]
I want to test in my chat application that an element is scrolled down to the bottom on the first load and keeps scrolled down on new messages. I'm scrolling down using the following code snippet:

`messagesElement.scrollTop = messagesElement.scrollHeight - messagesElement.offsetHeight;`

I've managed to test that the element is scrolled down on the first load with following code:

```
cy.mount(
  <Provider store={ store }>
    <App />
  </Provider>
);

cy.get('.message').its('length').should('eq', 20);

cy.get('.messages').then(el => {
  let e  = el.get(0);

  expect(e.scrollTop).to.eq(e.scrollHeight - e.offsetHeight);
});
```

But the second test with keeping element at the bottom **sometimes** fails. I'm using the following code:

```
cy.mount(
  <Provider store={ store }>
    <App />
  </Provider>
);

cy.get('.message').its('length').should('eq', 20);

let message: Message = {
  id: nanoid(),
  dialog: dialog.id,
  author: user.id,
  date: Math.floor((new Date()).getTime() / 1000),
  content: "Lorem ipsum dolor sit amet"
} as Message;

cy.window().its('store').invoke('dispatch', pushMessage(message));

cy.get('.messages').should('contain', message.content);

cy.wait(1000);

cy.get('.messages').then(el => {
  let e  = el.get(0);

  expect(e.scrollTop).to.eq(e.scrollHeight - e.offsetHeight);
});
```
As I understand it, cypress needs time to scroll the element down, but sometimes it doesn’t even enough 3 seconds to scroll down, and I don’t know how to write the test correctly. Also, using `wait(...)` method seems undoubtedly wrong.

How can I write such a test?

*In the second video, in which the test fails, if you look closely, you can see how the element did not scroll to the very bottom.

https://user-images.githubusercontent.com/6327801/222946960-de188582-ddba-445b-a21d-9b4e02d14101.mp4

https://user-images.githubusercontent.com/6327801/222946963-6ea8b4c4-b3a9-4599-9b3e-db0218497d0e.mp4

**Comment by portyu9** [comment]
Hi @someApprentice 

The main issue is that the scroll update is asynchronous, while the assertion inside `.then()` only runs once.

Instead of waiting one second and testing exact equality, put the calculation in a `.should()` callback so Cypress retries it until the UI settles:

```js
cy.get('.messages').should(($messages) => {
  const el = $messages[0]

  const distanceFromBottom =
    el.scrollHeight - el.clientHeight - el.scrollTop

  expect(Math.abs(distanceFromBottom)).to.be.lessThan(2)
})
```

This gives you two advantages:

- `.should()` automatically retries while the Redux/render/scroll update is occurring.
- comparing with a small tolerance avoids brittle exact-pixel equality caused by layout/scroll rounding.

You therefore shouldn't need:

```js
cy.wait(1000)
```

Also, for calculating the visible scrolling viewport, `clientHeight` is generally the appropriate companion to `scrollHeight`, rather than `offsetHeight`.

If your intended behavior is specifically “the messages container should remain at the bottom after a new message is rendered”, asserting the small `distanceFromBottom` value directly expresses that requirement.

Please mark this answer as accepted if it helped, thank you.
